File: java/messages/src/main/java/com/nedap/retail/messages/epcis/v1_1/ObjectEvent.java
package com.nedap.retail.messages.epcis.v1_1;

import java.util.List;
import java.util.Objects;

import org.codehaus.jackson.annotate.JsonProperty;
import org.joda.time.DateTime;

import com.nedap.retail.messages.epcis.v1_1.cbv.Action;
import com.nedap.retail.messages.epcis.v1_1.cbv.Disposition;
import com.nedap.retail.messages.epcis.v1_1.cbv.EventType;
import com.nedap.retail.messages.epcis.v1_1.elements.QuantityElement;

/**
 * Required parameters to construct an object_event are: - epcList - *or* quantityList
 */
public class ObjectEvent extends EpcisEvent {

    @JsonProperty("epc_list")
    public List<String> epcList;

    @JsonProperty("quantity_list")
    public List<QuantityElement> quantityList;

    public String ilmd;

    public ObjectEvent() {
        type = EventType.ObjectEvent;
    }

    /**
     * An object_event captures information about an event pertaining to one or more physical or digital objects
     * identified by instance-level (EPC) or class-level (EPC Class) identifiers. Most object_events are envisioned to
     * represent actual observations of objects, but strictly speaking it can be used for any event a Capturing
     * Application wants to assert about objects, including for example capturing the fact that an expected observation
     * failed to occur.
     *
     * @param id The ID that identifies this message uniquely to an organization
     * @param eventTime The date and time at which the EPCIS Capturing Applications asserts the event occurred
     * @param recordTime The date and time at which this event was recorded by the EPCIS Repository.
     * @param eventTimeZoneOffset The time zone offset in effect at the time and place the event occurred, expressed as
     *            an offset from UTC
     * @param action How this event relates to the lifecycle of the EPCs named in this event.
     * @param bizLocation The business location where the objects associated with the EPCs may be found, until
     *            contradicted by a subsequent event.
     * @param readPoint The read point at which the event took place.
     * @param disposition The business condition of the objects associated with the EPCs, presumed to hold true until
     *            contradicted by a subsequent event.
     * @param epcList An unordered list of one or more EPCs naming specific objects to which the event pertained.
     * @param quantityList An unordered list of one or more QuantityElements identifying (at the class level) objects to
     *            which the event pertained.
     */
    public ObjectEvent(final String id, final DateTime eventTime, final DateTime recordTime,
            final String eventTimeZoneOffset, final Action action, final String bizLocation, final String readPoint,
            final Disposition disposition, final List<String> epcList, final List<QuantityElement> quantityList) {
        this.id = id;
        this.eventTime = eventTime;
        this.recordTime = recordTime;
        this.eventTimeZoneOffset = eventTimeZoneOffset;
        this.type = EventType.ObjectEvent;
        this.action = action != null ? action.action() : null;
        this.bizLocation = bizLocation;
        this.readPoint = readPoint;
        this.disposition = disposition != null ? disposition.disposition() : Disposition.UNKNOWN.disposition();
        this.epcList = epcList;
        this.quantityList = quantityList;
    }

    @Override
    public String toString() {
        final String epcListSize = epcList == null ? "null" : Integer.toString(epcList.size());
        final String quantityListSize = quantityList == null ? "null" : Integer.toString(quantityList.size());

        return "ObjectEvent" + super.toString() + "[epcList(" + epcListSize + "),quantityList(" + quantityListSize
                + ")]";
    }

    @Override
    public boolean equals(final Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        if (!super.equals(obj)) {
            return false;
        }
        final ObjectEvent other = (ObjectEvent) obj;
        if (!compareAsSet(this.epcList, other.epcList)) {
            return false;
        }
        if (!compareAsSet(this.quantityList, other.quantityList)) {
            return false;
        }
        if (!Objects.equals(this.ilmd, other.ilmd)) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        result = 31 * result + (epcList != null ? epcList.hashCode() : 0);
        result = 31 * result + (quantityList != null ? quantityList.hashCode() : 0);
        result = 31 * result + (ilmd != null ? ilmd.hashCode() : 0);
        return result;
    }
}

File: csharp/NedapRetailApiClient/Nedap.Retail.Api/Epcis.V1_1/ObjectEvent.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using System.Collections.Generic;

namespace Nedap.Retail.Api.Epcis.V1_1
{
    /// <summary>
    /// An ObjectEvent captures information about an event pertaining to one or more physical or digital objects identified by instance-level (EPC) or class-level (EPC Class) identifiers. Most ObjectEvents are envisioned to represent actual observations of objects, but strictly speaking it can be used for any event a Capturing Application wants to assert about objects, including for example capturing the fact that an expected observation failed to occur.
    /// While more than one EPC may appear in an ObjectEvent, no relationship or association between those EPCs is implied other than the coincidence of having experienced identical events in the real world.The Action parameter of an ObjectEvent describes the event's relationship to the lifecycle of the objects and their identifiers named in the event.
    /// </summary>
    public class ObjectEvent : EpcisEvent
    {
        /// <summary>
        /// Constructor that sets type to object_event
        /// </summary>
        public ObjectEvent()
        {
            base.Type = "object_event";
        }

        /// <summary>
        /// How this event relates to the lifecycle of the EPCs named in this event.
        /// </summary>
        [JsonProperty("action")]
        [JsonConverter(typeof(StringEnumConverter))]
        public Action Action { get; set; }

        /// <summary>
        /// The business condition of the objects associated with the EPCs, presumed to hold true until contradicted by a subsequent event.
        /// </summary>
        [JsonProperty("disposition")]
        public string Disposition { get; set; }

        /// <summary>
        /// The business location where the objects associated with the EPCs may be found, until contradicted by a subsequent event.
        /// </summary>
        [JsonProperty("biz_location")]
        public string BizLocation { get; set; }

        /// <summary>
        /// An unordered list of one or more EPCs naming specific objects to which the event pertained. An ObjectEvent must contain either an epc_list or a quantity_list.
        /// </summary>
        [JsonProperty("epc_list")]
        public List<string> EpcList { get; set; }

        /// <summary>
        /// An unordered list of one or more QuantityElements identifying (at the class level) objects to which the event pertained. An ObjectEvent must contain either an epc_list or a quantity_list.
        /// </summary>
        [JsonProperty("quantity_list")]
        public List<QuantityElement> QuantityList { get; set; }

        /// <summary>
        /// Optional. The business step of which this event was a part.
        /// </summary>
        [JsonProperty("biz_step")]
        public string BizStep { get; set; }

        /// <summary>
        /// The read point at which the event took place.
        /// </summary>
        [JsonProperty("read_point")]
        public string ReadPoint { get; set; }

        /// <summary>
        /// Optional. An unordered list of business transactions that define the context of this event.
        /// </summary>
        [JsonProperty("biz_transaction_list")]
        public List<BusinessTransaction> BizTransactionList { get; set; }

        /// <summary>
        /// Optional. An unordered list of Source elements that provide context about the originating endpoint of a business transfer of which this event is a part.
        /// </summary>
        [JsonProperty("source_list")]
        public List<Source> SourceList { get; set; }

        /// <summary>
        /// Optional. An unordered list of Destination elements that provide context about the terminating endpoint of a business transfer of which this event is a part.
        /// </summary>
        [JsonProperty("destination_list")]
        public List<Destination> DestinationList { get; set; }

        /// <summary>
        /// Optional. Instance/Lot Master Data that describes the objects created during this event.
        /// </summary>
        [JsonProperty("ilmd")]
        public string Ilmd { get; set; }
    }
}